File: 622-design-circular-queue/622-design-circular-queue.cpp
class MyCircularQueue {
public:
    int front,rear,size,k;
    vector<int>arr;
    MyCircularQueue(int sz) {
        k=sz;
        arr.resize(k);
        front =-1,rear=-1;
        size=0;
    }
    
    bool enQueue(int value) {
        if(size==k) return false;
        if(front ==-1){
            front=0;
            rear=0;
            arr[front]=value;
            ++size;
            return true;
        }
        ++rear;
        rear%=k;
        arr[rear]=value;
        ++size;
        return true;
    }
    
    bool deQueue() {
        if(size==0) return false;
        front++;
        front%=k;
        --size;
        return true;
    }
    
    int Front() {
        if( size==0) return -1;
        return arr[front];
    }
    
    int Rear() {
        if( size==0) return -1;
        return arr[rear];
    }
    
    bool isEmpty() {
        return size==0;
    }
    
    bool isFull() {
        return size==k;
    }
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */

File: 622-design-circular-queue/622-design-circular-queue.py
class MyCircularQueue:

    def __init__(self, k: int):
        self.size=0
        self.maxsize=k
        self.t= [0]*k
        self.front=self.rear=-1

    def enQueue(self, value: int) -> bool:
        if self.size==self.maxsize:
            return False
        else :
            if self.rear==-1:
                self.rear=self.front=0
            else:
                self.rear= (self.rear+1)%self.maxsize
            self.t[self.rear]=value
            self.size+=1;
            return True

    def deQueue(self) -> bool:
        if self.size==0:
            return False
        if self.front==self.rear:
            self.front=self.rear=-1
        else :
            self.front=(self.front+1)%self.maxsize
        self.size-=1;
        return True
        

    def Front(self) -> int:
        if self.size==0:
            return -1
        else :
            return self.t[self.front]

    def Rear(self) -> int:
        if self.size==0:
            return -1
        else :
            return self.t[self.rear]
        

    def isEmpty(self) -> bool:
        return self.size==0
        

    def isFull(self) -> bool:
        return self.size==self.maxsize
        


# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()